src/include/function/cast/functions/cast_string_non_nested_functions.h

```diff
@@ -269,48 +276,61 @@ bool tryDecimalCast(const char* input, uint64_t len, T& result, uint32_t scale)
     bool negativeFlag = input[0] == '-';
     if (negativeFlag) {
         input++;
+        len -= 1;
     }
 
-    IntegerCastData<T> res{0};
+    CAST_DATA res;
+    res.result = 0;
     uint64_t pos = 0;
-    uint64_t periodPos = len;
+    uint64_t periodPos = len - 1;
     while (pos < len) {
         if (input[pos] == '.') {
             periodPos = pos;
-        }
-        if (!StringUtils::CharacterIsDigit(input[pos])) {
-            return false;
-        }
-        uint8_t digit = input[pos] - '0';
-        if (pos - periodPos > scale) {
-            // then determine rounding
-            if (digit >= 5) {
-                res.result += 1;
+        } else {
+            if (!StringUtils::CharacterIsDigit(input[pos])) {
```

Comment:
I would suggest refactor this block of code as a function

Reply:
The function may be equally messy; There's a conditional break that I'd have to work into a function setting, and I'd have to pass the CAST_OP, CAST_DATA templates, input, res, scale, pos, and periodPos parameters. 

However, I did manage to denest much of the code without putting it into a function.